src/ray/core_worker/store_provider/plasma_store_provider.cc

```diff
@@ -407,6 +407,9 @@ Status CoreWorkerPlasmaStoreProvider::Wait(
     if (check_signals_) {
       RAY_RETURN_NOT_OK(check_signals_());
     }
+
+    // Adjust num_objects to request based on the number of objects already found.
+    num_objects -= static_cast<int>(ready->size());
```

Comment:
would changing the above condition to something like
```
if (ready.size() == static_cast<size_t>(num_objects)) {
    should_break = true;
}
```
after everything from `result_pair.first` is added to ready also fix your issue and limit side-effects for the request the raylet_client makes

Or is the idea that the request also has an inflated num_objects so changing that also helps there

Reply:
The change would likely need to be `(ready.size() >= static_cast<size_t>(num_objects))` since you could still end up waiting for more objects than requested. I think its better to decrease the number of objects requested in each iteration to avoid requesting more than the user wanted

Reply:
This in a for loop to request to wait objects from plasma store. In the first round there are some objects ready but it's not enough. In the second round, *the same `id_vector` is requested* and if we reduce num_objects we may hit false positives for the objects made ready in the first round.

So if we are going to subtract from the num_objects shall we also subtract read object IDs from  `id_vector`?

I think current code (no changes to id_vector or num_objects) should be semantically correct, since we just ask for the already-ready objects again and it's only a matter of performance. I will think more about why the CHECK fails.